src/rdkafka_timer.c

```diff
@@ -195,18 +196,36 @@ void rd_kafka_timer_start0(rd_kafka_timers_t *rkts,
 
         rd_kafka_timers_unlock(rkts);
 }
-
 /**
  * Delay the next timer invocation by '2 * rtmr->rtmr_interval'
+ * @param minimum_backoff the minimum backoff to be applied
+ * @param maximum_backoff the maximum backoff to be applied
+ * @param max_jitter the jitter percentage to be applied to the backoff
  */
 void rd_kafka_timer_exp_backoff(rd_kafka_timers_t *rkts,
-                                rd_kafka_timer_t *rtmr) {
+                                rd_kafka_timer_t *rtmr,
+                                rd_ts_t minimum_backoff,
+                                rd_ts_t maximum_backoff,
+                                int max_jitter) {
+        int64_t actual_interval;
         rd_kafka_timers_lock(rkts);
         if (rd_kafka_timer_scheduled(rtmr)) {
-                rtmr->rtmr_interval *= 2;
                 rd_kafka_timer_unschedule(rkts, rtmr);
         }
-        rd_kafka_timer_schedule(rkts, rtmr, 0);
+
+        rtmr->rtmr_interval *= 2;
+        if (rtmr->rtmr_interval < minimum_backoff)
+                rtmr->rtmr_interval = minimum_backoff;
+        if ((rtmr->rtmr_interval > maximum_backoff) && (maximum_backoff != -1))
```

Comment:
If you remember the rtmr interval tells the exact backoff at which we are capped by retry max ms. I am just maininting that invariant. I know it would still work without the if but it is good to maintain.

Reply:
There are some cases it will be wrong without the if.
Consider one case, where retry.backoff.ms = 100, retry.max.ms = 1000, and we are on retries = 6.

In this case, 
rtmr interval = 2**(6-1)*100 = 3200
rtmr interval = 1000 as it exceeds max

assume jitter is +10%, so actual backoff becomes 1100. We reset the actual backoff to the max again, equal to 1000. 

All good.

Then the request fails again, we are on retries = 7.
 rtmr interval = 2**(6-1)*100 = 6400
rtmr interval = 1000 as it exceeds max

assume jitter is -10%, so actual backoff becomes 900. We don't reset the actual backoff as it is below 1000.

So the backoff is ... 1000 -> 900.
The backoff decreases between attempts.

That's not what the KIP stipulates. Once it hits 1000, it has to stay 1000.

There's maybe a potential problem with my suggested solution, because of the subtraction actual_backoff - rtmr interval below. Maybe this approach suggested here will work - https://github.com/confluentinc/librdkafka/pull/4422#discussion_r1328390278

I hope I've illustrated why it's wrong. Please fix it. 

Please don't resolve comments without addressing them or if you have started a discussion on them.

Reply:
Have made the requested change!